test/cmd/lfstest-gitserver.go

```diff
@@ -0,0 +1,235 @@
+		gitHandler(w, r)
+	})
+
+	urlname := os.Getenv("LFSTEST_URL")
+	if len(urlname) == 0 {
+		urlname = "lfstest-gitserver"
+	}
+
+	file, err := os.Create(urlname)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	file.Write([]byte(server.URL))
+	file.Close()
+	log.Println(server.URL)
+
+	defer func() {
+		os.RemoveAll(urlname)
+	}()
+
+	<-stopch
+	log.Println("git server done")
+}
+
+type lfsObject struct {
+	Oid   string             `json:"oid,omitempty"`
+	Size  int64              `json:"size,omitempty"`
+	Links map[string]lfsLink `json:"_links,omitempty"`
+}
+
+type lfsLink struct {
+	Href   string            `json:"href"`
+	Header map[string]string `json:"header,omitempty"`
+}
+
+// handles any requests with "{name}.server.git/info/lfs" in the path
+func lfsHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/vnd.git-lfs+json")
+	if r.Method == "POST" {
+		lfsPostHandler(w, r)
+	} else {
+		lfsGetHandler(w, r)
+	}
+}
+
+func lfsPostHandler(w http.ResponseWriter, r *http.Request) {
+	buf := &bytes.Buffer{}
+	tee := io.TeeReader(r.Body, buf)
+	obj := &lfsObject{}
+	err := json.NewDecoder(tee).Decode(obj)
+	io.Copy(ioutil.Discard, r.Body)
+	r.Body.Close()
+
+	log.Println("REQUEST")
+	log.Println(buf.String())
+	log.Printf("OID: %s\n", obj.Oid)
+	log.Printf("Size: %d\n", obj.Size)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	res := &lfsObject{
+		Links: map[string]lfsLink{
+			"upload": lfsLink{
+				Href: server.URL + "/storage/" + obj.Oid,
+			},
+		},
+	}
+
+	by, err := json.Marshal(res)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Println("RESPONSE: 202")
+	log.Println(string(by))
+
+	w.WriteHeader(202)
+	w.Write(by)
+}
+
+func lfsGetHandler(w http.ResponseWriter, r *http.Request) {
+	parts := strings.Split(r.URL.Path, "/")
+	oid := parts[len(parts)-1]
+
+	if by, ok := largeObjects[oid]; ok {
+		obj := &lfsObject{
+			Oid:  oid,
+			Size: int64(len(by)),
+			Links: map[string]lfsLink{
+				"download": lfsLink{
+					Href: server.URL + "/storage/" + oid,
+				},
+			},
+		}
+
+		by, err := json.Marshal(obj)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		log.Println("RESPONSE: 200")
+		log.Println(string(by))
+
+		w.WriteHeader(200)
+		w.Write(by)
+
+		return
+	}
+
+	w.WriteHeader(404)
+}
+
+// handles any /storage/{oid} requests
+func storageHandler(w http.ResponseWriter, r *http.Request) {
+	log.Printf("storage %s %s\n", r.Method, r.URL)
+	switch r.Method {
+	case "PUT":
+		hash := sha256.New()
+		buf := &bytes.Buffer{}
+		io.Copy(io.MultiWriter(hash, buf), r.Body)
+		oid := hex.EncodeToString(hash.Sum(nil))
+		if !strings.HasSuffix(r.URL.Path, "/"+oid) {
+			w.WriteHeader(403)
+			return
+		}
+
+		largeObjects[oid] = buf.Bytes()
+
+	case "GET":
+		parts := strings.Split(r.URL.Path, "/")
+		oid := parts[len(parts)-1]
+
+		if by, ok := largeObjects[oid]; ok {
+			w.Write(by)
+			return
+		}
+
+		w.WriteHeader(404)
+	default:
+		w.WriteHeader(500)
```

Comment:
Honestly, because it's a test server and I don't really care.  Though I suppose the first time I hit that bug in testing, I'll curse myself for not using something more descriptive like 405 :)